File: ios/RecipeApp/ViewModels/AuthViewModel.swift
import Foundation
import Combine

@MainActor
class AuthViewModel: ObservableObject {
    @Published var isAuthenticated = false
    @Published var currentUser: User?
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let authService = AuthService.shared

    init() {
        checkAuthentication()
    }

    func checkAuthentication() {
        isAuthenticated = authService.isAuthenticated()
        if isAuthenticated {
            Task {
                await loadProfile()
            }
        }
    }

    func login(email: String, password: String) async {
        isLoading = true
        errorMessage = nil

        do {
            let response = try await authService.login(email: email, password: password)
            currentUser = response.user
            isAuthenticated = true
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    func register(email: String, password: String, username: String) async {
        isLoading = true
        errorMessage = nil

        do {
            let response = try await authService.register(email: email, password: password, username: username)
            currentUser = response.user
            isAuthenticated = true
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    func logout() async {
        isLoading = true

        do {
            try await authService.logout()
            currentUser = nil
            isAuthenticated = false
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    private func loadProfile() async {
        do {
            currentUser = try await authService.getProfile()
        } catch {
            // If profile load fails, user might need to log in again
            isAuthenticated = false
            currentUser = nil
        }
    }
}

File: android/app/src/main/java/com/recipemanager/viewmodel/AuthViewModel.kt
package com.recipemanager.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.recipemanager.data.model.User
import com.recipemanager.data.repository.RecipeRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed class AuthState {
    object Idle : AuthState()
    object Loading : AuthState()
    data class Success(val user: User) : AuthState()
    data class Error(val message: String) : AuthState()
}

class AuthViewModel(
    private val repository: RecipeRepository = RecipeRepository()
) : ViewModel() {

    private val _authState = MutableStateFlow<AuthState>(AuthState.Idle)
    val authState: StateFlow<AuthState> = _authState.asStateFlow()

    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated.asStateFlow()

    init {
        checkAuthentication()
    }

    private fun checkAuthentication() {
        viewModelScope.launch {
            _isAuthenticated.value = repository.isAuthenticated()
        }
    }

    fun login(email: String, password: String) {
        viewModelScope.launch {
            _authState.value = AuthState.Loading

            val result = repository.login(email, password)

            result.fold(
                onSuccess = { response ->
                    _authState.value = AuthState.Success(response.user)
                    _isAuthenticated.value = true
                },
                onFailure = { error ->
                    _authState.value = AuthState.Error(
                        error.message ?: "Login failed. Please try again."
                    )
                }
            )
        }
    }

    fun register(email: String, password: String, username: String) {
        viewModelScope.launch {
            _authState.value = AuthState.Loading

            val result = repository.register(email, password, username)

            result.fold(
                onSuccess = { response ->
                    _authState.value = AuthState.Success(response.user)
                    _isAuthenticated.value = true
                },
                onFailure = { error ->
                    _authState.value = AuthState.Error(
                        error.message ?: "Registration failed. Please try again."
                    )
                }
            )
        }
    }

    fun logout() {
        viewModelScope.launch {
            repository.logout()
            _isAuthenticated.value = false
            _authState.value = AuthState.Idle
        }
    }

    fun resetAuthState() {
        _authState.value = AuthState.Idle
    }
}
